src/java/org/apache/cassandra/index/sai/disk/v4/V4OnDiskFormat.java

```diff
@@ -0,0 +1,73 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.cassandra.index.sai.disk.v4;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+
+import org.apache.cassandra.index.sai.IndexContext;
+import org.apache.cassandra.index.sai.SSTableContext;
+import org.apache.cassandra.index.sai.disk.v1.IndexSearcher;
+import org.apache.cassandra.index.sai.disk.v1.PerIndexFiles;
+import org.apache.cassandra.index.sai.disk.v1.SegmentMetadata;
+import org.apache.cassandra.index.sai.disk.v3.V3OnDiskFormat;
+import org.apache.cassandra.index.sai.utils.TypeUtil;
+import org.apache.cassandra.utils.bytecomparable.ByteComparable;
+import org.apache.cassandra.utils.bytecomparable.ByteSource;
+import org.apache.cassandra.utils.bytecomparable.ByteSourceInverse;
+
+public class V4OnDiskFormat extends V3OnDiskFormat
+{
+    public static final V4OnDiskFormat instance = new V4OnDiskFormat();
+
+    @Override
+    public IndexSearcher newIndexSearcher(SSTableContext sstableContext,
+                                          IndexContext indexContext,
+                                          PerIndexFiles indexFiles,
+                                          SegmentMetadata segmentMetadata) throws IOException
+    {
+        if (indexContext.isLiteral())
+            return new V4InvertedIndexSearcher(sstableContext.primaryKeyMapFactory(), indexFiles, segmentMetadata, sstableContext.indexDescriptor, indexContext);
+        return super.newIndexSearcher(sstableContext, indexContext, indexFiles, segmentMetadata);
+    }
+
+    @Override
+    public boolean trieRangeRequiresValueValidation()
+    {
+        // Now that we encode the trie values correctly, we do not need to test the range.
+        return false;
+    }
+
+    @Override
+    public ByteComparable encode(ByteBuffer input, IndexContext indexContext)
+    {
+        // Composite values are considered literal, except for their encoding.
+        if (indexContext.isLiteral() && !TypeUtil.isComposite(indexContext.getValidator()))
+            return version -> append(ByteSource.of(input, version), ByteSource.TERMINATOR);
+        return version -> TypeUtil.asComparableBytes(input, indexContext.getValidator(), version);
```

Comment:
These are all very good questions. I wondered the same about using the `AbstractType` to encode the byte comparable. I don't know the historical reason that we haven't, but I am thinking we might be able to use `TypeUtil.asComparableBytes(input, indexContext.getValidator(), version);` after all. The main outstanding question is whether we need to have the `ByteSource.TERMINATOR` as the suffix. I left a comment in the most recent commit.

Reply:
For example, here is a UTF8 encoding:

```
unencoded:                   [76, 76, 83, 121, 67, 65, 80, 67, 98, 88, 71, 72, 77, 71, 90, 85, 86, 105, 67, 110, 75, 100, 105, 75, 122, 102]
type encoded:                [76, 76, 83, 121, 67, 65, 80, 67, 98, 88, 71, 72, 77, 71, 90, 85, 86, 105, 67, 110, 75, 100, 105, 75, 122, 102, 0]
type encoded with terimator: [76, 76, 83, 121, 67, 65, 80, 67, 98, 88, 71, 72, 77, 71, 90, 85, 86, 105, 67, 110, 75, 100, 105, 75, 122, 102, 0, 56]
```

As you can see, the type encoding adds `0` to the end (it also escapes all 0s), and then the terminator adds the 56 to the end of that.

Reply:
A couple of notes on the original `ByteComparable` encodings and why SAI may have chosen to use something different:
- The translations of types in `AbstractType.asComparableBytes` are primarily meant to be used inside multi-component clustering keys. As such they need to be composable into a flat key, which can be achieved by making them prefix-free (actually [weakly prefix free](https://github.com/datastax/cassandra/blob/main/src/java/org/apache/cassandra/utils/bytecomparable/ByteComparable.md#desired-properties)) at the expense of making some of them longer.
- This type encoding is also useful to address the points before and after an entry (using a different terminator value), making it easier to handle inclusive/exclusive bounds in the trie code for range queries (the tries _can_ currently give precise inclusive/exclusive slices, but this may be removed soon as it's difficult to define for reverse cursors).
- In the primary index code, `ByteComparable` implies prefix-free, and `ByteSource`: weakly prefix-free. The naming of methods in the two interfaces is based on this understanding. Neither of these is typically true for SAI's uses.
- SAI indexes single values, thus the encoding being prefix-free is not essential.
- There was an expectation that pattern-matching (e.g. "abc*") was to be implemented very soon after the very first SAI version, for which storing strings directly was more useful.
- However, if we store strings/blobs directly without escaping, we can't add a terminator, because this will violate the ordering of a prefix entry compared to some of its extensions.

From my point of view it makes sense to move SAI in line with what the primary index code does (i.e. always use `AbstractType`'s encoding with an added terminator to make `ByteComparable`s prefix-free) to make things easier to follow. There is a performance cost to that, and a complication for pattern matching, which should both be relatively minor.

Reply:
Thank you for this context @blambov. I am good with switching to make SAI follow the primary index code.

For reference, I attempted to use the `AbstractType`'s encoding for all values in the trie with this commit https://github.com/datastax/cassandra/pull/1095/commits/9657d4d7421d42e650f6023f26af67fca70d20af on this branch, but I reverted it because there were test failures. I agree that it would be simpler if we had the same encoding, so I'll see if I can address those test failures easily. If we can't, I might defer that work because the framework added in this PR should make it easier to migrate later.